### PR TITLE
feat: SC-19444 cell highlighting

### PR DESCRIPTION
**Pull request by JonnCharpentier**
Feature request from [ticket](https://app.shortcut.com/homebound-team/story/19444/add-cell-highlighting-to-bulk-schedule-phases)

Add cell highlighting 

This branch is based off https://github.com/homebound-team/beam/pull/626
image

### CODE REVIEW

File: src/components/Table/GridTable.tsx
Comment (stephenh) [inline src/components/Table/GridTable.tsx:1129]: Why the `useComputed`? I suppose it's b/c we're accessing `rowState.activeCellId`...

In theory `GridTable` will also do this if `GridTable.observeRows` is set (i.e. wrap `GridRow` in an `ObservableGridRow`)...

Maybe we should change:

```
      const RowComponent = observeRows ? ObservedGridRow : MemoizedGridRow;
```

To be `observeRows || cellHighlight ? ... :`

Or we could just _always_ use `ObservedGridRow`, it probably wouldn't hurt anything, and we could remove the `observeRows` prop all together (b/c it's always set true now); maybe just do that?

File: src/components/Table/GridTable.tsx
Comment (stephenh) [reply src/components/Table/GridTable.tsx:254]: (I suppose if page really wanted to set it, they could use the `GridTableApi`.)

File: src/components/Table/GridTable.tsx
Comment (JonnCharpentier) [reply src/components/Table/GridTable.tsx:1129]: by doing this, removing the observeRows prop, are we introducing a breaking change on internal-frontend?

Comment (JonnCharpentier) [reply src/components/Table/GridTable.tsx:1129]: @bdow what do you think?

File: src/components/Table/GridTable.tsx
Comment (JonnCharpentier) [reply src/components/Table/GridTable.tsx:254]: yeah, in this case the active cell is not only highlighted, but also open up a side pane that display task details

File: src/components/Table/GridTable.tsx
Comment (bdow) [reply src/components/Table/GridTable.tsx:1129]: I kind of like the idea of just _always_ doing `observeRows`. Maybe for now keep the prop, but default it to be `true`. This should not be a breaking change on BP.

Comment (bdow) [reply src/components/Table/GridTable.tsx:1129]: I'm also good with removing the prop too... Would just mean we need to clean up places that set it, which would be an easy fix.

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:97]: **request:** I'm thinking I would prefer this to be an option on the `GridStyleDef` in `styles.tsx`, and not here. If that is set, then we can add in hover styles for the `cellCss` defined in the `memoizedTableStyles` function. 
The reasoning for this request is because we want to move away from people using `GridStyle` and instead choose a `GridStyleDef`. This would ensure better consistency between all of our tables.

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1241]: When we add the `hover` styles to GridStylesDef, then we can remove it here.

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1239]: **question:** Does this cause all the cells to shift when we add in this border? This could impact how we define our column widths and potentially cause horizontal scrollbars when a table is already the full width of the page 🤔. It might be fine since we are using `box-sizing: border-box`, but it might cause the content within the cell to shift around. We might want to add a transparent border to every cell to prevent any shifting.

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1438]: **request:** This isn't your fault, but it really confused me when reading this. We **aren't** rendering a `Row` here, but instead a `Cell`. This naming and the use of a `tr` tag is incorrect in this situation. Would you mind fixing this and updating it to:
```
const Cell = as === "table" ? "td" : "div";
return (
  <Cell>...</Cell>
);
```
Please and thanks!

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:344]: **question:** Did we get rid of `MemoizedGridRow`? I don't think its being used still is it?

File: src/components/Table/GridTable.test.tsx
Comment (bdow) [inline src/components/Table/GridTable.test.tsx:2079]: **question:** Does this fail now? I'm assuming so since you skipped it. 
If so, @stephenh do you think this is a bad choice to remove the `MemoizedGridRow` if `ObservedGridRow` will cause rerenders like this?

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1225]: **nit:** You do not need the `else` block here. You could _always_ set `bTransparent` and only override it with the `if` statement.

File: src/components/Table/GridTable.test.tsx
Comment (bdow) [inline src/components/Table/GridTable.test.tsx:2124]: **question:** Shouldn't the other cells have the border style, but with a transparent color?

File: src/components/Table/GridTable.test.tsx
Comment (bdow) [inline src/components/Table/GridTable.test.tsx:2125]: I really like the test for the style rules. Can we write tests to ensure (1) The active cell is updated if the prop value changes, and (2) the active cell can be changed via the GridTableApi?

File: src/components/Table/GridTable.stories.tsx
Comment (bdow) [inline src/components/Table/GridTable.stories.tsx:107]: **question:** Is this necessary? Won't the `cellOnClick ` in the GridRow do this for us?

File: src/components/Table/GridTable.tsx
Comment (stephenh) [reply src/components/Table/GridTable.tsx:344]: Ah damn; @bdow / @JonnCharpentier it looks like using `ObservedGridRow` means that (even before Jonn's PR) we would just never use `MemoizedGridRow`.

I think this was probably a bug/mistake in the old code, b/c the idea was less of "choose either observed or memoized" (which is what it was doing) but the intention was "observed would be layered on top of memoized".

@JonnCharpentier I wonder if we can try:

```
const MemoizedGridRow = React.memo(observer(GridRow), (one, two) => {
  const { row: row1, ...others1 } = one;
  const { row: row2, ...others2 } = two;
  return shallowEqual(row1, row2) && shallowEqual(others1, others2);
}) as typeof GridRow;
```

And then make line 354 `<MemoizedGridRow ...`.

I believe this will achieve both goals, of memoizing the component based on the `row` prop, but also wrapping it in an observer (now using the out-of-the-box `observer` function instead of the previous hand-written `ObservedGridRow` which was just doing the same thing).

File: src/components/Table/GridTable.stories.tsx
Comment (JonnCharpentier) [reply src/components/Table/GridTable.stories.tsx:107]: on this story I'm doing both active cell and active Row this onClick is just for setting active row

Comment (stephenh) [reply src/components/Table/GridTable.stories.tsx:107]: :thinking: should we have this line do both:

```
const cellOnClick = applyCellHighlight ? () => api.setActiveCellId(cellId) : undefined;
```

Set the active cell id as well as the active row id?

My initial thought is that this seems great, b/c it starts to move control of "what is the active row?" into GridTable, instead of being something each page needs to know about / reimplement.

Comment (JonnCharpentier) [reply src/components/Table/GridTable.stories.tsx:107]: Y es, we can do that

Comment (JonnCharpentier) [reply src/components/Table/GridTable.stories.tsx:107]: @stephenh what if we only want to highlight a cell, but don't want to use active row styles? I believe this is the case on my ticket

Comment (bdow) [reply src/components/Table/GridTable.stories.tsx:107]: Sorry, I misread that code, I thought you were setting the active cell in that `onClick` callback. I'm good with them being treated separately.

Comment (stephenh) [reply src/components/Table/GridTable.stories.tsx:107]: @JonnCharpentier ah I see, merely setting the active row id turns on active row styles?

I wonder if we need some sort of `rowHighlight`, similar to `cellHighlight`, that tables need to set to explicitly turn on row highlighting behavior...

We can probably deal with that later if you'd like to leave the current `cellOnClick` as-is.

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1081]: **nit:** `rowId` should now be part of the dependency array.

**Review by bdow — COMMENTED** [review]
Looks like a lot of storybook diff issues with how the borders are applied now. Appears as though it is an issue with how the row borders are applied, I think via `betweenRowsCss` in the `export const defaultStyle: GridStyle`. I _think_ they way these styles are defined is done on each cell element, vs the row element. We initially did this due to the row element being `display: contents` and couldn't have styles applied to it. This is no longer the case, so we should be able to apply the `betweenRowCss` on the row elements themselves.

File: src/components/Table/GridTable.tsx
Comment (JonnCharpentier) [reply src/components/Table/GridTable.tsx:1081]: question, does'nt  putting rowId on that dependency array will make it fire useComputed every render? since is a variable that get re-defined on every render, also, if we already have row object as a dependency, then if the row changes we will get the new rowId, and rowId can not change unless row changed

**Comment by JonnCharpentier** [comment]
> is is no longer the case, so we should be able to apply the `betweenRowCss` on the row elements themselves.

I don't understand what the request is here

**Comment by bdow** [comment]
> > is is no longer the case, so we should be able to apply the `betweenRowCss` on the row elements themselves.
> 
> I don't understand what the request is here

There are a bunch of GridTable stories that look very off due to borders being applied around the _whole_ cell, except the cells in the first row ([current](https://main--60466f7d43c37c0021788867.chromatic.com/?path=/story/components-gridtable--client-side-sorting) vs [this PR](https://60466f7d43c37c0021788867-rgothkhvbb.chromatic.com/?path=/story/components-gridtable--client-side-sorting)). This is a result of the new cell border styles in this PR and how the `betweenRowCss` is applied are not compatible. I am thinking the `betweenRowCss` is being applied to each cell element, and causing this.

**Review by bdow — COMMENTED** [review]
Looks like there are still some issues with the borders showing up when they shouldn't in the Storybook diffs: 
https://www.chromatic.com/test?appId=60466f7d43c37c0021788867&id=6320b324f201b76a50f5dd9f

**Comment by bdow** [comment]
@JonnCharpentier Another thought I had, which I _think_ I mentioned in our meeting, is what if we used a `box-shadow` for the active cell styling? Then we wouldn't need to set any border styles, transparent or with a color. That might help solve some of this.
```
box-shadow: inset 0 0 0 2px Palette.LightBlue900;
```

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:549]: Let's add ` > *` back in to the selector, but leave the rest as is.

File: src/components/Table/GridTable.tsx
Comment (bdow) [reply src/components/Table/GridTable.tsx:1081]: Since `rowId` is a string, it will be properly evaluated as `"rowId" === "rowId"`. There should be no difference, and would not cause the `useComputed` to fire again.

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1226]: Maybe need the `br4` here?

File: src/components/Table/GridTable.tsx
Comment (bdow) [inline src/components/Table/GridTable.tsx:1413]: Let's remove the `onClick` from the `rowLinkRenderFn`.

File: src/components/Table/styles.tsx
Comment (bdow) [inline src/components/Table/styles.tsx:75]: Remove this.

File: src/components/Table/styles.tsx
Comment (bdow) [inline src/components/Table/styles.tsx:77]: Add box-shadow back

**Review by bdow — APPROVED** [review]
As far as I can tell, all the snapshots look right. I'll write this up somewhere, but I tend to find it better to not approve storybook diffs until after the code has been approved. That way you're usually comparing against `main`.

**Comment by JonnCharpentier** [comment]
> As far as I can tell, all the snapshots look right. I'll write this up somewhere, but I tend to find it better to not approve storybook diffs until after the code has been approved. That way you're usually comparing against `main`.

I didn't know that's how it works,  Thanks! I'll do a beam version update

**Pull request merged**
